exporter/sumologicexporter/sender.go

```diff
@@ -0,0 +1,258 @@
+	// sent gives information if the data was sent or not
+	sent bool
+	// appended keeps state of appending new log line to the body
+	appended bool
+	err      error
+}
+
+type sender struct {
+	buffer []pdata.LogRecord
+	config *Config
+	client *http.Client
+	filter filter
+	ctx    context.Context
+}
+
+const (
+	logKey string = "log"
+	// maxBufferSize defines size of the buffer (maximum number of pdata.LogRecord entries)
+	maxBufferSize int = 1024 * 1024
+)
+
+func newAppendResponse() appendResponse {
+	return appendResponse{
+		appended: true,
+	}
+}
+
+func newSender(ctx context.Context, cfg *Config, cl *http.Client, f filter) *sender {
+	return &sender{
+		config: cfg,
+		client: cl,
+		filter: f,
+		ctx:    ctx,
+	}
+}
+
+// send sends data to sumologic
+func (s *sender) send(pipeline PipelineType, body io.Reader, flds fields) error {
+	// Add headers
+	req, err := http.NewRequestWithContext(s.ctx, http.MethodPost, s.config.HTTPClientSettings.Endpoint, body)
+	if err != nil {
+		return err
+	}
+
+	req.Header.Add("X-Sumo-Client", s.config.Client)
+
+	if len(s.config.SourceHost) > 0 {
+		req.Header.Add("X-Sumo-Host", s.config.SourceHost)
+	}
+
+	if len(s.config.SourceName) > 0 {
+		req.Header.Add("X-Sumo-Name", s.config.SourceName)
+	}
+
+	if len(s.config.SourceCategory) > 0 {
+		req.Header.Add("X-Sumo-Category", s.config.SourceCategory)
+	}
+
+	switch pipeline {
+	case LogsPipeline:
+		req.Header.Add("Content-Type", "application/x-www-form-urlencoded")
+		req.Header.Add("X-Sumo-Fields", string(flds))
+	case MetricsPipeline:
+		// ToDo: Implement metrics pipeline
+		return errors.New("current sender version doesn't support metrics")
+	default:
+		return errors.New("unexpected pipeline")
+	}
+
+	resp, err := s.client.Do(req)
+	if err != nil {
+		return err
+	}
+	if resp.StatusCode < 200 || resp.StatusCode >= 400 {
+		return fmt.Errorf("error during sending data: %s", resp.Status)
+	}
+	return nil
+}
+
+// logToText converts LogRecord to a plain text line, returns it and error eventually
+func (s *sender) logToText(record pdata.LogRecord) (string, error) {
+	return record.Body().StringVal(), nil
+}
+
+// logToText converts LogRecord to a json line, returns it and error eventually
+func (s *sender) logToJSON(record pdata.LogRecord) (string, error) {
+	data := s.filter.filterOut(record.Attributes())
+	data[logKey] = record.Body().StringVal()
+
+	nextLine, err := json.Marshal(data)
+	if err != nil {
+		return "", err
+	}
+
+	return bytes.NewBuffer(nextLine).String(), nil
+}
+
+// sendLogs sends log records from the buffer formatted according
+// to configured LogFormat and as the result of execution
+// returns array of records which has not been sent correctly and error
+func (s *sender) sendLogs(flds fields) ([]pdata.LogRecord, error) {
+	var (
+		body           strings.Builder
+		errs           []error
+		droppedRecords []pdata.LogRecord
+		currentRecords []pdata.LogRecord
+		formattedLine  string
+		err            error
```

Comment:
The same comment applies to other properties, but do you need `err` to be at this scope? You are one bug away from a problem that is hard to detect :-) Move the `err` inside the loop, so that you won't leak it elsewhere.

Reply:
Moved together with `formattedLine`. Other variables require broaden scope